### PR TITLE
Foundry ci setup

### PR DESCRIPTION
**Pull request by r0qs**
This PR extract the Foundry/python CI part from https://github.com/ethereum/solidity/pull/13873 to simplify the review.

~Depends on https://github.com/ethereum/solidity/pull/14064~

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -93,6 +93,75 @@ commands:
             - bytecode-report-<< parameters.label >>-cli.txt
       - matrix_notify_failure_unless_pr
 
+  install_python3:
+    description: "Install python3 and given packages."
+    parameters:
+      packages:
+        description: "List of optional python packages to be installed (separated by space)."
```
Comment (cameel) [inline .circleci/config.yml]:
```suggestion
        description: "List of extra Python packages to be installed (separated by space)."
```

File: .circleci/config.yml [hunk 2]
```diff
@@ -705,14 +774,8 @@ jobs:
     <<: *base_cimg_small
```
Comment (cameel) [inline .circleci/config.yml:767]: Would be best if we could avoid having to install Python and pip every time and just have it as a part of the image. Maybe we should switch `chk_pylint`, `chk_proofs` and `c_ext_benchmarks` to `base_python_small` instead?

Though I guess we still won't avoid it completely - for Foundry we want rust image as a base. We could also leave it like this in this PR and refactor separately.

File: .circleci/config.yml [hunk 3]
```diff
@@ -93,6 +93,75 @@ commands:
...
+    steps:
+      - run:
+          name: Install python and dependencies
+          command: |
+            sudo apt update
+            if ! python3 --version >/dev/null 2>&1
+            then
+              sudo apt install python3 --assume-yes --no-install-recommends
+            fi
+            sudo apt install python3-pip
```
Comment (cameel) [inline .circleci/config.yml]: Was unconditional installation causing a problem in some cases where Python is already installed? We generally don't mind getting newest Python so this would be fine too if it works:

```suggestion
            sudo apt install python3 python3-pip --assume-yes --no-install-recommends
```

File: .circleci/config.yml [hunk 4]
```diff
@@ -93,6 +93,75 @@ commands:
...
+            if [ -n "<< parameters.packages >>" ]
```
Comment (cameel) [inline .circleci/config.yml]:
```suggestion
            if [[ "<< parameters.packages >>" != "" ]]
```

File: .circleci/config.yml [hunk 5]
```diff
@@ -1256,17 +1318,17 @@ jobs:
             equal: [<< parameters.binary_type >>, "solcjs"]
           steps:
             - run:
-                name: External << parameters.project >> tests (solcjs)
+                name: External << parameters.project >> tests (<< parameters.binary_type >>)
                 command: |
-                  test/externalTests/<< parameters.project >>.sh solcjs /tmp/workspace/soljson.js
+                  test/external_tests.py test --solc-binary-type << parameters.binary_type >> --solc-binary-path /tmp/workspace/soljson.js --run << parameters.project >>
```
Comment (cameel) [inline .circleci/config.yml]: I'd quote shell values that come from variables as a general rule, even if currently it does not seem necessary:
```suggestion
                  test/external_tests.py test --solc-binary-type "<< parameters.binary_type >>" --solc-binary-path /tmp/workspace/soljson.js --run "<< parameters.project >>"
```

It also makes it look a bit less like you're doing input/output redirection :)

File: .circleci/config.yml [hunk 6]
```diff
@@ -93,6 +93,75 @@ commands:
...
+
+  install_foundry:
+    description: "Install Foundry."
+    parameters:
+      version:
+        type: string
+        default: "nightly"
+      install_path:
+        type: string
+        default: /home/circleci/bin
```
Comment (cameel) [inline .circleci/config.yml:123]: Why not install it system-wide in `/usr/local/bin`?

```suggestion
        default: /usr/local/bin/
```

File: .circleci/config.yml [hunk 7]
```diff
@@ -93,6 +93,75 @@ commands:
...
+    steps:
+      - run:
+          name: Setup Foundry environment variables
+          command: |
+            FOUNDRY_REPO="foundry-rs/foundry"
+            FOUNDRY_VERSION=<< parameters.version >>
```
Comment (cameel) [inline .circleci/config.yml]:
```suggestion
            FOUNDRY_VERSION="<< parameters.version >>"
```

File: .circleci/config.yml [hunk 8]
```diff
@@ -93,6 +93,75 @@ commands:
...
+              --fail \
+              --show-error \
+              "https://api.github.com/repos/${FOUNDRY_REPO}/git/refs/tags/${FOUNDRY_VERSION}" \
+              | jq --raw-output .object.sha \
+            )
+            echo "export FOUNDRY_REPO=$FOUNDRY_REPO" >> "$BASH_ENV"
+            echo "export FOUNDRY_VERSION=$FOUNDRY_VERSION" >> "$BASH_ENV"
+            echo "export FOUNDRY_RELEASE_TAG='nightly-${FOUNDRY_RELEASE_SHA}'" >> "$BASH_ENV"
+            # Save commit sha for caching
+            echo $FOUNDRY_RELEASE_SHA > /tmp/workspace/foundry-release-sha
```
Comment (cameel) [inline .circleci/config.yml]:
```suggestion
            echo "$FOUNDRY_RELEASE_SHA" > /tmp/workspace/foundry-release-sha
```

File: .circleci/config.yml [hunk 9]
```diff
@@ -93,6 +93,75 @@ commands:
...
+      - restore_cache:
+          keys:
+            - foundry-<< parameters.version >>-{{ checksum "/tmp/workspace/foundry-release-sha" }}
+      - run:
+          name: Install foundry
+          command: |
+            if ! command -v forge > /dev/null 2>&1; then
+              wget \
+                --output-document /tmp/foundry.tar.gz \
+                "https://github.com/${FOUNDRY_REPO}/releases/download/${FOUNDRY_RELEASE_TAG}/foundry_${FOUNDRY_VERSION}_linux_amd64.tar.gz"
```
Comment (cameel) [inline .circleci/config.yml]: It's not like it wrong, but it's kinda weird to randomly switch between `curl` and `wget`. Why not stick to one?

```suggestion
              curl \
                --output /tmp/foundry.tar.gz \
                "https://github.com/${FOUNDRY_REPO}/releases/download/${FOUNDRY_RELEASE_TAG}/foundry_${FOUNDRY_VERSION}_linux_amd64.tar.gz"
```

File: .circleci/config.yml [hunk 10]
```diff
@@ -93,6 +93,75 @@ commands:
...
+              sudo tar --extract --gzip --file /tmp/foundry.tar.gz --directory << parameters.install_path >>
+            else
+              echo "Foundry/forge version: $(forge --version | cut -f2 -d' ') already installed. Skipping installation..."
+            fi
+      - save_cache:
+          key: foundry-<< parameters.version >>-{{ checksum "/tmp/workspace/foundry-release-sha" }}
+          paths:
+            - << parameters.install_path >>
```
Comment (cameel) [inline .circleci/config.yml]: We usually add a warning like this, because it's easy to forget and unintentionally put something here:

https://github.com/ethereum/solidity/blob/56e7352691dddecfa2e1faebdb4a8a6a11c5a8a2/.circleci/config.yml#L1418-L1419

File: .circleci/config.yml [hunk 11]
```diff
@@ -93,6 +93,75 @@ commands:
...
+          keys:
+            - foundry-<< parameters.version >>-{{ checksum "/tmp/workspace/foundry-release-sha" }}
+      - run:
+          name: Install foundry
+          command: |
+            if ! command -v forge > /dev/null 2>&1; then
+              wget \
+                --output-document /tmp/foundry.tar.gz \
+                "https://github.com/${FOUNDRY_REPO}/releases/download/${FOUNDRY_RELEASE_TAG}/foundry_${FOUNDRY_VERSION}_linux_amd64.tar.gz"
+              sudo tar --extract --gzip --file /tmp/foundry.tar.gz --directory << parameters.install_path >>
```
Comment (cameel) [inline .circleci/config.yml]: Instead of unpacking the whole tar directly into a dir that's in our `PATH`, perhaps it would be safer to unpack it to a subdir and symlink only the 4 binaries we expect to be there?
```suggestion
              cd "<< parameters.install_path >>"
              sudo tar --extract --gzip --file /tmp/foundry.tar.gz --directory foundry/
              sudo ln --symbolic foundry/{forge,anvil,cast,chisel} .
```

Otherwise, if foundry gets compromised and someone puts out a malicious tar, it can shadow any binary in the system with one it contains. It's not like I expect tars from foundry to be particularly unsafe, but better to have defense in depth.

File: .circleci/config.yml [hunk 12]
```diff
@@ -93,6 +93,75 @@ commands:
...
+            echo "export FOUNDRY_RELEASE_TAG='nightly-${FOUNDRY_RELEASE_SHA}'" >> "$BASH_ENV"
+            # Save commit sha for caching
+            echo $FOUNDRY_RELEASE_SHA > /tmp/workspace/foundry-release-sha
+      - restore_cache:
+          keys:
+            - foundry-<< parameters.version >>-{{ checksum "/tmp/workspace/foundry-release-sha" }}
+      - run:
+          name: Install foundry
+          command: |
+            if ! command -v forge > /dev/null 2>&1; then
```
Comment (cameel) [inline .circleci/config.yml]: This step is never supposed to run if the cache gets restored, so I think we can assume foundry is not installed.

Or you could leave just a small sanity check in case we switch images and this turns out to be wrong:
```suggestion
            ! forge --version 2> /dev/null
```

File: .circleci/config.yml [hunk 13]
```diff
@@ -93,6 +93,75 @@ commands:
...
+              | jq --raw-output .object.sha \
+            )
+            echo "export FOUNDRY_REPO=$FOUNDRY_REPO" >> "$BASH_ENV"
+            echo "export FOUNDRY_VERSION=$FOUNDRY_VERSION" >> "$BASH_ENV"
+            echo "export FOUNDRY_RELEASE_TAG='nightly-${FOUNDRY_RELEASE_SHA}'" >> "$BASH_ENV"
+            # Save commit sha for caching
+            echo $FOUNDRY_RELEASE_SHA > /tmp/workspace/foundry-release-sha
+      - restore_cache:
+          keys:
+            - foundry-<< parameters.version >>-{{ checksum "/tmp/workspace/foundry-release-sha" }}
```
Comment (cameel) [inline .circleci/config.yml]: It would be better to include the platform in the cache key:

```suggestion
            - foundry-<< parameters.version >>-{{ arch }}-{{ checksum "/tmp/workspace/foundry-release-sha" }}
```

Comment (cameel) [inline .circleci/config.yml]: You have the hash in an env var, you should be able to avoid storing it in a file and just use it directly according to [`save_cache` docs](https://circleci.com/docs/configuration-reference/#savecache):

```suggestion
            - foundry-<< parameters.version >>-{{ .Environment.FOUNDRY_RELEASE_SHA }}
```

File: .circleci/config.yml [hunk 14]
```diff
@@ -1256,17 +1318,17 @@ jobs:
...
       - when:
           condition:
             equal: [<< parameters.binary_type >>, "native"]
           steps:
             - run:
-                name: External << parameters.project >> tests (native)
+                name: External << parameters.project >> tests (<< parameters.binary_type >>)
                 command: |
-                  test/externalTests/<< parameters.project >>.sh native /tmp/workspace/solc/solc-static-linux
+                  test/external_tests.py test --solc-binary-type << parameters.binary_type >> --solc-binary-path /tmp/workspace/solc/solc-static-linux --run << parameters.project >>
```
Comment (cameel) [inline .circleci/config.yml]: If you extract this bit into a separate PR, we could merge it right away. It does not depend on Foundry in any way, only on the the external PR refactor, which is now merged, and the only remark I had here were the missing quotes.

File: .circleci/config.yml (hunk 2 above)
Comment (r0qs) [reply .circleci/config.yml:767]: Yes, we could do the switch but probably in another PR as you suggested.

File: .circleci/config.yml (hunk 6 above)
Comment (r0qs) [reply .circleci/config.yml:123]: I did like that at first, but as it is an external script and foundry currently does not provide a checksum for their releases, I thought that it would be better to avoid installing things using `sudo`. But just realized now that I left a `sudo` there when extracting the binaries. Just fix it ;)

File: .circleci/config.yml (hunk 9 above)
Comment (r0qs) [reply .circleci/config.yml]: No big reason, I usually use `wget` to download files from the web for better error reporting and useful defaults. For instance, in the way you suggested, `curl` would download an empty file (because it does not follow redirections by default) or in case of errors, it would download a text file containing an error string, while `wget` would immediately report `404`. Of course this can be fixed by adding other parameters to `curl` like `--fail --location` but yeah, I changed to `curl` :)

File: .circleci/config.yml (hunk 11 above)
Comment (r0qs) [reply .circleci/config.yml]: True. Yeah, it would be nice if they provide checksum for their releases.

File: .circleci/config.yml (hunk 13 above)
Comment (r0qs) [reply .circleci/config.yml]: Hum...I couldn't make it work in the way you suggested. If I understand correctly their docs, this syntax does not include environment variables that defined in our CI scripts, only the ones that are added in the `context` of the project/organization CircleCI settings or that are already exported by CircleCI.

File: .circleci/config.yml (hunk 3 above)
Comment (r0qs) [reply .circleci/config.yml]: No, I was just trying to avoid installing it if it was already in the image, since the command is used by different images and I didn't want to risk break other dependencies.

File: .circleci/config.yml (hunk 14 above)
Comment (r0qs) [reply .circleci/config.yml]: Done here: https://github.com/ethereum/solidity/pull/14191

**Review by ekpyron — APPROVED** [review]
